codecov.yml

```diff
@@ -1,15 +1,49 @@
 #
-# For documentation see https://codecov.io/gh/twisted/twisted/settings/yaml
+# For documentation: https://docs.codecov.io/docs/codecovyml-reference
+# Twisted settings: https://codecov.io/gh/twisted/twisted/settings/yaml
 #
 # We want 100% coverage for new patches to make sure we are always increasing
 # the coverage.
 #
+codecov:
+  require_ci_to_pass: yes
+  notify:
+    # We have at least 10 builds in GitHub Actions and 12 in Azure
+    # and lint + mypy + docs + ReadTheDocs
+    after_n_builds: 15
+    wait_for_ci: yes
+
 coverage:
   precision: 2
   round: down
   status:
     patch:
       default:
+        # New code should have 100% CI coverage as the minimum
+        # quality assurance measurement.
+        # If there is a good reason for new code not to have coverage,
+        # add inline pragma comments.
         target: 100%
+    project:
+      default:
+        # Allow for a bit of slack in overall project report.
+        # As long as the patch is 100% the PR should be ok.
+        # We report coverage even on errors.
+        # So a run on trunk might fail and during the failure process will
+        # exercise a different code branch (ex timeout errors).
+        # This code branch is reported at the end of the run.
+        # Future test runs will not trigger the error and the error handling
+        # code branch is not called, and if no other tests is touching that
+        # branch we are left with overall reduced code coverage.
+        threshold: 0.02%
```

Comment:
This setting may be required if coverage is a required check but the reasoning here is invalid.  The threshold value allows an ongoing slip in coverage.  Build N of trunk can have (0.02 × M) greater coverage than build (N+M) of trunk and this configuration tells codecov.io that this is fine.

So this setting should be considered a temporary fix while the real problem is being addressed - non-deterministic coverage in the test suite.

Reply:
There are various cases where coverage can be reduced.  Even just deleting lines that have been covered causes this.  It's just a result of not having '100%' coverage (after including consideration of coverage ignores).  I don't know if there's a way to deal with that separately from the commented situation.

Reply:
What are the odds, that is the case in my next review.  https://github.com/twisted/towncrier/pull/334  They deleted a line that is useless and the coverage dropped by 0.01%.

Reply:
Thanks wiml for the review. Just monitoring future PR and checking for overall project coverage regression can help with future fixed for this issue.

Thanks exarkun for the headup. I forgot to file the followup ticket.

Thanks altendky for the code delete info. I was hoping that codeco.io is smart and will not trigger such an error. 
I have reported this issue 
https://community.codecov.io/t/project-coverage-decrease-on-deleted-line/2726

-----------------

I agree that this should be a temporary change. 
I have added to link to the ticket in the comment.

If coverage/project will continue to be red, people will most probably ignore it by default.
With that, I think that is a higher probability of seeing an overall decrease in coverage.

We had the project red for some time, and I am not aware of any secondary tickets or actions to fix the overall project coverage.

-----------

Regarding 0.02 x M.
We have 100% patch coverage , so is not that easy to find a series of PR  with 100% patch coverage but witch will always decreased the overall project by 0.02 due to different tests.

I am monitoring overall coverage changes  as there is a gang of usual suspect tests.

For example

![image](https://user-images.githubusercontent.com/204609/113612848-4424b480-9648-11eb-958f-2376fcc37f94.png)

With this swinging line

![image](https://user-images.githubusercontent.com/204609/113613671-75ea4b00-9649-11eb-9247-545efd2807bc.png)

There is also this swinging test helpers to prevent a flaky test failure

![image](https://user-images.githubusercontent.com/204609/113614262-2fe1b700-964a-11eb-88d0-dd3505d0978a.png)